### PR TITLE
Exit early if there are no matrix entries to add.

### PR DESCRIPTION
**Pull request by drwells**
Believe it or not but this is related to #2325.

This gets around an issue where, if one is using a sparsity pattern created by `make_flux_sparsity_pattern` and the `nonzero` coupling option, one may end up trying to add zeros to entries that do not exist in the sparsity pattern. This problem manifests itself when adding fluxes between locally owned and ghost cells. The problem seems to be with Trilinos, not deal.II (I suppose it does not elide zeros). This commit sidesteps the issue by just not calling Trilinos at all if there is nothing (i.e., the only thing to do is add zero) to do.

### CODE REVIEW

**Comment by kronbichler** [comment]
Just to help me understanding the problem: There is nothing to add for a particular row on a given processor because there we no entries added to the row because of `Coupling::nonzero`. So Trilinos ends up having a matrix row of length zero (it is most likely in the off-processor matrix that we hold). What is the error message that comes out of it? I'm just curious to understand whether it is really Epetra's fault or the way we use it. Is it because the row length is zero (but the number of added elements should be zero, too)?

Apart from that, I completely agree with the patch. The test is very nice. Would you mind adding a changelog entry?

**Comment by drwells** [comment]
I made a small improvement to the test: there are two cells so it should be run with `mpirun=2`, not `mpirun=4`. I also added the changelog entry.

The underlying issue is something like this (and I updated the test with this information): suppose we have the linear advection equation (unfortunately we have to work in 2D here) `u_t = u_x + u_y` (1). Then, for a two cell triangulation

```
 2 ------- 3 6 ------- 7
 |         | |         |
 |    0    | |    1    |
 |         | |         |
 0 ------- 1 4 ------- 5
```

with `FE_DGQ(1)` elements there should be no coupling between the degrees of freedom on the zeroth cell and basis functions 5 and 7 on the first cell with most numerical fluxes since basis functions 5 and 7 are zero on the 1-3 edge. Hence the sparsity pattern on rank 0 does not contain a (5, 5) or (7, 7) entry.

This is troublesome with MeshWorker or any DG solver that tries to integrate across each face *exactly once*: the `neighbor_to_neighbor_flux` matrix across the 1-3 face could contain all zeros (and will certainly be zero for rows or columns corresponding to basis functions 5 and 7). Hence I currently get the following message when running this test on `master` in debug mode:

```
2534: --------------------------------------------------------
2534: An error occurred in line <1678> of file
<.../source/lac/trilinos_sparse_matrix.cc> in function
2534:     void dealii::TrilinosWrappers::SparseMatrix::add
(dealii::TrilinosWrappers::SparseMatrix::size_type, dealii::TrilinosWrappers::
SparseMatrix::size_type, const size_type*, const TrilinosScalar*, bool, bool)
2534: The violated condition was:
2534:     nonlocal_matrix->RowMap().LID(static_cast<TrilinosWrappers::types::int_type>(row)) != -1
2534: The name and call sequence of the exception was:
2534:     ExcMessage("Attempted to write into off-processor matrix row " "that
has not be specified as being writable upon " "initialization")
2534: Additional Information:
2534: Attempted to write into off-processor matrix row that has not be specified
as being writable upon initialization
2534: --------------------------------------------------------
```

What Trilinos does seems wrong but I am certainly not an expert in the library: even when `n_columns == 0` it triggers an exception, so it seems like Trilinos does something here to reach an undesired state.

(1): personal note: there are some descendents of Heinz Kreiss here at RPI and the habit of using this equation to solve everything is rubbing off on me.

**Comment by kronbichler** [comment]
@drwells This exception is our fault, not Trilinos' fault. The design of the TrilinosWrappers::SparseMatrix class uses two matrices, where the second one is called `nonlocal_matrix` and only used for handling matrix entries that should be handled by another processor during assembly. (This is quite a lot of code on our side, but when you talk to Trilinos developers this is the 'official' way of handling finite element problems in the more modern way, rather than Epetra_FECrsMatrix which has always remained a niche and is not supported in Tpetra for example.)

What we do is that we initialize the `nonlocal_matrix` from the non-owned rows of the sparsity pattern, but only those that have some entries to contribute (see lines 657-658 of trilinos_sparse_matrix.cc). But then in line 1678 where you get the exception we do not check this any more.

Bottom line: Thanks for the patch, this is exactly the right thing to do and must be part of our wrapper class.

> (1): personal note: there are some descendents of Heinz Kreiss here at RPI and the habit of using this equation to solve everything is rubbing off on me.

Having done my PhD at an institute that was founded by Heinz Kreiss (Scientific Computing in Uppsala) I know what you are talking about. Well, I did my PhD for Heinz' daughter Gunilla, which means that my primary (academic) descendants are actually not on his line but the one of Guermund Dahlquist, the other famous Swedish from the good old time of computing :-)

File: tests/trilinos/elide_zeros.cc [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+ * 1-3 edge. Hence the sparsity pattern on rank 0 does not contain a (5, 5) or
+ * (7, 7) entry.
+ *
+ * This is troublesome with MeshWorker or any DG solver that tries to
+ * integrate across each face *exactly once*: the neighbor_to_neighbor_flux
+ * matrix across the 1-3 face could contain all zeros (and will certainly be
+ * zero for rows or columns corresponding to basis functions 5 and 7). Hence I
+ * currently get the following message when running this test on master in
+ * debug mode:
+ *
```
Comment (kronbichler) [inline tests/trilinos/elide_zeros.cc]: Would you mind changing the text in the last sentence to something like: "This would previously give the following exception:"

File: tests/trilinos/elide_zeros.cc [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+ * > |    0    | |    1    |
+ * > |         | |         |
+ * > 0 ------- 1 4 ------- 5
+ *
+ * with FE_DGQ(1) elements there should be no coupling between the degrees
+ * of freedom on the zeroth cell and basis functions 5 and 7 on the first cell
+ * with most numerical fluxes since basis functions 5 and 7 are zero on the
+ * 1-3 edge. Hence the sparsity pattern on rank 0 does not contain a (5, 5) or
+ * (7, 7) entry.
+ *
```
Comment (kronbichler) [inline tests/trilinos/elide_zeros.cc]: Very nice description of the test case :+1:

**Comment by kronbichler** [comment]
When done with this minor change, please merge yourself.

**Comment by drwells** [comment]
@kronbichler Small world: the people here where at Caltech with him.

I updated the commit messages to make them clearer and remove my wrong accusation of Trilinos.
